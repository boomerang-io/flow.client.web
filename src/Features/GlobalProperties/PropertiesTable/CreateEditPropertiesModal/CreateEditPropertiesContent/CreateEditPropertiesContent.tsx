import React from "react";
import { useMutation, useQueryClient } from "react-query";
import { Formik } from "formik";
import * as Yup from "yup";
import {
  Button,
  InlineNotification,
  Loading,
  ModalBody,
  ModalFlowForm,
  ModalFooter,
  notify,
  ToastNotification,
  TextInput,
  Toggle,
} from "@boomerang-io/carbon-addons-boomerang-react";
import { InputType, PROPERTY_KEY_REGEX, PASSWORD_CONSTANT } from "Constants";
import { serviceUrl, resolver } from "Config/servicesConfig";
import { Property, PatchProperty } from "Types";
import { updatedDiff } from "deep-object-diff";

type Props = {
  closeModal: () => void;
  isEdit: boolean;
  property: Property;
  propertyKeys: string[];
  cancelRequestRef: any;
};

const configUrl = serviceUrl.getGlobalConfiguration();

function CreateEditPropertiesContent({ closeModal, isEdit, property, propertyKeys, cancelRequestRef }: Props) {
  const initialState = {
    label: property?.label ?? "",
    description: property?.description ?? "",
    key: property?.key ?? "",
    value: property?.value ?? "",
    secured: property?.type === InputType.Secured ?? false,
  };

  const queryClient = useQueryClient();
  /** Add property */
  const { mutateAsync: addGlobalPropertyMutation, isLoading: addLoading, error: addError } = useMutation(
    (args: { body: Property }) => {
      const { promise, cancel } = resolver.postGlobalPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryClient.invalidateQueries(configUrl),
    }
  );

  /** Update property */
  const { mutateAsync: updateGlobalPropertyMutation, isLoading: updateLoading, error: updateError } = useMutation(
    (args: { id: string; body: PatchProperty }) => {
      const { promise, cancel } = resolver.patchGlobalPropertyRequest(args);
      cancelRequestRef.current = cancel;
      return promise;
    },
    {
      onSuccess: () => queryClient.invalidateQueries(configUrl),
    }
  );

  const loading = addLoading || updateLoading;

  const handleSubmit = async (values: any) => {
    const type = values.secured ? InputType.Secured : InputType.Text;
    const newProperty = isEdit ? { ...values, type, id: property.id } : { ...values, type };
    delete newProperty.secured;

    if (isEdit) {
      const updatedFields = updatedDiff(initialState, newProperty);
      try {
        await updateGlobalPropertyMutation({ id: newProperty.id, body: updatedFields });
        notify(
          <ToastNotification
            kind="success"
            title={"Property Updated"}
            subtitle={`Request to update ${newProperty.label} succeeded`}
            data-testid="create-update-global-prop-notification"
          />
        );
        closeModal();
      } catch (err) {}
    } else {
      try {
        await addGlobalPropertyMutation({ body: newProperty });
        notify(
          <ToastNotification
            kind="success"
            title="Property Created"
            subtitle="Request to create property succeeded"
            data-testid="create-update-global-prop-notification"
          />
        );
        closeModal();
      } catch (err) {}
    }
  };

  // Check if key contains alpahanumeric, underscore, dash, and period chars
  const validateKey = (key: any) => {
    return PROPERTY_KEY_REGEX.test(key);
  };

  return (
    <Formik
      initialValues={initialState}
      onSubmit={handleSubmit}
      validateOnMount
      validationSchema={Yup.object().shape({
        label: Yup.string().required("Enter a label"),
        key: Yup.string()
          .required("Enter a key")
          .max(128, "Key must not be greater than 128 characters")
          .notOneOf(propertyKeys || [], "Enter a unique key value for this property")
          .test(
            "is-valid-key",
            "Only alphanumeric, hyphen and underscore characters allowed. Must begin with a letter or underscore",
            validateKey
          ),
        value: Yup.string().when("secured", {
          is: (value: boolean) => value && isEdit,
          then: Yup.string(),
          otherwise: Yup.string().required("Enter a value"),
        }),
        description: Yup.string(),
        secured: Yup.boolean(),
      })}
    >
      {(props) => {
        const { values, touched, errors, isValid, handleChange, handleBlur, handleSubmit } = props;

        return (
          <ModalFlowForm onSubmit={handleSubmit}>
            {loading && <Loading />}
            <ModalBody aria-label="inputs">
              <TextInput
                data-testid="create-parameter-key"
                id="key"
                labelText="Key"
                name="key"
                value={values.key}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.key && touched.key}
                invalidText={errors.key}
              />
              <TextInput
                data-testid="create-parameter-label"
                id="label"
                labelText="Label"
                name="label"
                value={values.label}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.label && touched.label}
                invalidText={errors.label}
              />
              <TextInput
                data-testid="create-parameter-description"
                id="description"
                labelText="Description (optional)"
                name="description"
                value={values.description}
                onBlur={handleBlur}
                onChange={handleChange}
              />
              <TextInput
                data-testid="create-parameter-value"
                id="value"
                labelText="Value"
                placeholder={isEdit && values.secured ? PASSWORD_CONSTANT : "Value"}
                name="value"
                value={values.value}
                onBlur={handleBlur}
                onChange={handleChange}
                invalid={errors.value && touched.value}
                invalidText={errors.value}
                type={values.secured ? "password" : "text"}
                helperText={
                  isEdit && values.secured
                    ? "Default values are stored securely and cannot be viewed once created. To update, provide a new default value."
                    : null
                }
              />
              <Toggle
                id="secured-global-parameters-toggle"
                disabled={isEdit}
                labelText="Secured"
                name="secured"
                onChange={handleChange}
                orientation="vertical"
                toggled={values.secured}
                data-testid="secured-global-parameters-toggle"
                helperText="Once a parameter is securely created - you will not be able to view it or make it unsecure."
              />

              {addError && (
                <InlineNotification
                  lowContrast
                  kind="error"
                  title={"Create Property Failed"}
                  subtitle={"Something's Wrong"}
                  data-testid="create-update-global-prop-notification"
                />
              )}
              {updateError && (
                <InlineNotification
                  lowContrast
                  kind="error"
                  title={"Update Property Failed"}
                  subtitle={"Something's Wrong"}
                  data-testid="create-update-global-prop-notification"
                />
              )}
            </ModalBody>
            <ModalFooter>
              <Button kind="secondary" type="button" onClick={closeModal}>
                Cancel
              </Button>
              <Button
                type="submit"
                disabled={!isValid || loading}
                data-testid="global-parameter-create-submission-button"
              >
                {isEdit ? (loading ? "Saving..." : "Save") : loading ? "Creating..." : "Create"}
              </Button>
            </ModalFooter>
          </ModalFlowForm>
        );
      }}
    </Formik>
  );
}

export default CreateEditPropertiesContent;
