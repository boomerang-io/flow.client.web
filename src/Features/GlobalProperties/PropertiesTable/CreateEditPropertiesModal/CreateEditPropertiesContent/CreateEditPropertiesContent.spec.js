import React from "react";
import CreateEditPropertiesContent from ".";

const mockfn = jest.fn();
const props = {
  isEdit: true,
  property: { id: "1", key: "test.key", value: "testing" },
  propertyKeys: ["test"],
  handleEditClose: mockfn,
  addPropertyInStore: mockfn,
  updatePropertyInStore: mockfn,
  cancelRequestRef: { current: null },
};

describe("CreateEditPropertiesContent --- Snapshot Test", () => {
  it("Capturing Snapshot of CreateEditPropertiesContent", () => {
    const { baseElement } = global.rtlQueryRender(<CreateEditPropertiesContent {...props} />);
    expect(baseElement).toMatchSnapshot();
  });
});
