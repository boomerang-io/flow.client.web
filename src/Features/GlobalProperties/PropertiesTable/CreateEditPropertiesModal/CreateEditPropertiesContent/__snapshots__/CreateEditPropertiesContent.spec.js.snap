// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateEditPropertiesContent --- Snapshot Test Capturing Snapshot of CreateEditPropertiesContent 1`] = `
<body
  id="app"
>
  <div>
    <form
      class="bx--bmrg-modal-flow-form"
    >
      <div
        aria-label="inputs"
        class="bx--modal-content"
      >
        <div
          class="bx--bmrg-text-input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="key"
            >
              <div
                class="bx--bmrg-text-input__label"
              >
                <div>
                  Key
                </div>
              </div>
            </label>
            <div
              class="bx--text-input__field-outer-wrapper"
            >
              <div
                class="bx--text-input__field-wrapper"
              >
                <input
                  class="bx--text-input bx--text-input--md"
                  data-testid="create-parameter-key"
                  id="key"
                  name="key"
                  type="text"
                  value="test.key"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bx--bmrg-text-input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="label"
            >
              <div
                class="bx--bmrg-text-input__label"
              >
                <div>
                  Label
                </div>
              </div>
            </label>
            <div
              class="bx--text-input__field-outer-wrapper"
            >
              <div
                class="bx--text-input__field-wrapper"
              >
                <input
                  class="bx--text-input bx--text-input--md"
                  data-testid="create-parameter-label"
                  id="label"
                  name="label"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bx--bmrg-text-input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="description"
            >
              <div
                class="bx--bmrg-text-input__label"
              >
                <div>
                  Description (optional)
                </div>
              </div>
            </label>
            <div
              class="bx--text-input__field-outer-wrapper"
            >
              <div
                class="bx--text-input__field-wrapper"
              >
                <input
                  class="bx--text-input bx--text-input--md"
                  data-testid="create-parameter-description"
                  id="description"
                  name="description"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bx--bmrg-text-input"
        >
          <div
            class="bx--form-item bx--text-input-wrapper"
          >
            <label
              class="bx--label"
              for="value"
            >
              <div
                class="bx--bmrg-text-input__label"
              >
                <div>
                  Value
                </div>
              </div>
            </label>
            <div
              class="bx--text-input__field-outer-wrapper"
            >
              <div
                class="bx--text-input__field-wrapper"
              >
                <input
                  class="bx--text-input bx--text-input--md"
                  data-testid="create-parameter-value"
                  id="value"
                  name="value"
                  placeholder="Value"
                  title="Value"
                  type="text"
                  value="testing"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="bx--bmrg-toggle__container"
        >
          <div
            class="bx--bmrg-toggle --vertical"
          >
            <div
              class="bx--bmrg-toggle__title"
            >
              <label
                class="bx--label"
                for="secured-global-parameters-toggle"
                id="secured-global-parameters-toggle-label"
                style="margin-bottom: 0px;"
              >
                Secured
              </label>
            </div>
            <div
              class="bx--form-item"
            >
              <input
                aria-labelledby="secured-global-parameters-toggle-label"
                class="bx--toggle-input"
                data-testid="secured-global-parameters-toggle"
                disabled=""
                id="secured-global-parameters-toggle"
                name="secured"
                type="checkbox"
              />
              <label
                aria-label="Toggle"
                class="bx--toggle-input__label"
                for="secured-global-parameters-toggle"
              >
                <span
                  class="bx--toggle__switch"
                >
                  <span
                    aria-hidden="true"
                    class="bx--toggle__text--off"
                  />
                  <span
                    aria-hidden="true"
                    class="bx--toggle__text--on"
                  />
                </span>
              </label>
            </div>
          </div>
          <div
            class="bx--form__helper-text"
            style="margin-bottom: 0px;"
          >
            Once a parameter is created, secured state will not be able to be updated
          </div>
        </div>
      </div>
      <div
        class="bx--modal-footer bx--btn-set"
      >
        <button
          class="bx--btn bx--btn--secondary"
          tabindex="0"
          type="button"
        >
          Cancel
        </button>
        <button
          class="bx--btn bx--btn--primary"
          data-testid="global-parameter-create-submission-button"
          tabindex="0"
          type="submit"
        >
          Save
        </button>
      </div>
    </form>
  </div>
</body>
`;
